sympy/physics/mechanics/tests/test_joint.py

```diff
@@ -0,0 +1,113 @@
+from sympy.testing.pytest import raises
+
+t = dynamicsymbols._t
+
+def test_Joint():
+    parent = Body('parent')
+    child = Body('child')
+    raises(TypeError, lambda: Joint('J', parent, child))
+
+
+def test_pinjoint():
+    P = Body('P')
+    C = Body('C')
+    Pj = PinJoint('J', P, C)
+    theta, omega = dynamicsymbols('J_theta, J_omega')
+
+    assert Pj._name == 'J'
+    assert Pj.parent() == P
+    assert Pj.child() == C
+    assert Pj.coordinates() == [theta]
+    assert Pj.speeds() == [omega]
+    assert Pj.kdes() == [omega - theta.diff(t)]
+    assert Pj._parent_axis == P.frame.x
+    assert Pj._child_joint.pos_from(C.masscenter) == Vector(0)
+    assert Pj._parent_joint.pos_from(P.masscenter) == Vector(0)
+    assert Pj._parent_joint.pos_from(Pj._child_joint) == Vector(0)
+
+    l, m = symbols('l m')
+    J1 = PinJoint('J1', P, C, parent_joint_pos= l * P.frame.x, child_joint_pos= m * C.frame.y,
+        parent_axis = P.frame.z)
+
+    assert J1._parent_axis == P.frame.z
+    assert J1._child_joint.pos_from(C.masscenter) == m * C.frame.y
+    assert J1._parent_joint.pos_from(P.masscenter) == l * P.frame.x
+    assert J1._parent_joint.pos_from(J1._child_joint) == Vector(0)
+
+def test_pin_joint_double_pendulum():
+    q1, q2 = dynamicsymbols('q1 q2')
+    u1, u2 = dynamicsymbols('u1 u2')
+    m, l = symbols('m l')
+    N = ReferenceFrame('N')
+    A = ReferenceFrame('A')
+    B = ReferenceFrame('B')
+    O = Point('O')
+    P = O.locatenew('P', l * A.x)
+    R = P.locatenew('R', l * B.x)
+    C = Body('C', O, frame=N) #ceiling
+    PartP = Body('P', P, frame=A, mass=m)
+    PartR = Body('R', R, frame=B, mass=m)
+
+    J1 = PinJoint('J1', C, PartP, speeds=u1, coordinates=q1, parent_axis=C.frame.z)
+    J2 = PinJoint('J2', PartP, PartR, speeds=u2, coordinates=q2, parent_axis=PartP.frame.z)
+
+    #Check orientation
+    assert N.dcm(A) == Matrix([[cos(q1), -sin(q1), 0], [sin(q1), cos(q1), 0], [0, 0, 1]])
+    assert A.dcm(B) == Matrix([[cos(q2), -sin(q2), 0], [sin(q2), cos(q2), 0], [0, 0, 1]])
+    assert N.dcm(B).simplify() == Matrix([[cos(q1 + q2), -sin(q1 + q2), 0], \
+        [sin(q1 + q2), cos(q1 + q2), 0], [0, 0, 1]])
+
+    #Check Angular Velocity
+    assert A.ang_vel_in(N) == u1 * N.z
+    assert B.ang_vel_in(A) == u2 * A.z
+    assert B.ang_vel_in(N) == u1 * N.z + u2 * A.z
+
+    #Check kde
+    assert J1.kdes() == [u1 - q1.diff(t)]
+    assert J2.kdes() == [u2 - q2.diff(t)]
+
+    #Check Linear Velocity
+    assert PartP.masscenter.vel(N) == l*u1*A.y
+    assert PartR.masscenter.vel(A) == l*u2*B.y
+
+def test_pin_joint_chaos_pendulum():
+    mA, mB, lA, lB, h = symbols('mA, mB, lA, lB, h')
+    theta, phi, omega, alpha = dynamicsymbols('theta phi omega alpha')
+    N = ReferenceFrame('N')
+    A = ReferenceFrame('A')
+    B = ReferenceFrame('B')
+    No = Point('No')
+    Ao = Point('Ao')
+    Bo = Point('Bo')
+    lA = (lB - h / 2) / 2
+    Ao.set_pos(No, lA * A.z)
+    Bo.set_pos(No, lB * A.z)
```

Comment:
All positions should be set when the joints are formed. For example Ao and Bo wrt to No would happen when you form a PinJoint(A, N) and PinJoint(B, A). The joints method should relieve the user from having to figure out global positions.

Reply:
These are bodies mass centers and mass centers are needed to be provided to bodies if one wants to form Rigid Body

Reply:
Center of masses are need to be defined when the `Body` is being instantiated.

Reply:
We should be able to specify a Body with a Point that has no position relative to No. The idea is that each body has a local coordinate system with its origin at its mass center (as well as the "ground"). The Joint should establish relative positions and orientations to the two bodies and populate that info to the underlying RigidBody.

Reply:
But for that  wouldn't we need to take distance from user? 

Something like:

```
PinJoint('P', Par, Child, dist_btw_centerofmass=6*N.x)
```